stringfuzz/generators/equality.py

```diff
@@ -0,0 +1,107 @@
+import random
+
+from stringfuzz.scanner import ALPHABET
+from stringfuzz.smt import *
+
+__all__ = [
+    'equality',
+]
+
+def random_string(size):
+    return ''.join(random.choice(ALPHABET) for i in range(size))
+
+def get_length(max_length, randomise):
+    if randomise is False:
+        return max_length
+    return random.randint(0, max_length)
+
+def concat_terms(terms):
+
+    # initialise result to the last term (i.e. first in reversed list)
+    reversed_terms = reversed(terms)
+    result         = next(reversed_terms)
+
+    # keep concatenating preceding terms to the result
+    for term in reversed_terms:
+        result = smt_concat(term, result)
+
+    return result
+
+def randomly_add_infix(probability):
+    return random.random() < probability
+
+def make_equality(num_expressions, num_terms, prefix_length, suffix_length, add_infixes, infix_length, randomise_lengths, infix_probability):
+
+    # check args
+    if num_expressions < 1:
+        raise ValueError('the number of expressions must be at least 1')
+
+    if num_terms < 2:
+        raise ValueError('the number of terms per expression must be at least 2')
+
+    if infix_probability < 0.0 or 1.0 < infix_probability:
+        raise ValueError('the probability of infixes must be between 0.0 and 1.0')
+
+    # result values
+    expressions = []
+    variables   = []
+    heads       = []
+
+    # create expressions
+    for i in range(num_expressions):
+
+        # create head, prefix, and suffix
+        head   = smt_new_var()
```

Comment:
head, and prefix are quite similar. Since it ends up being head = prefix . middle . suffix maybe you can name them left = right_prefix . right_middle . right_suffix? Or some nicer variation

Reply:
I'm thinking of making there be only one main variable, and calling it `root`. What do you think of that?